Handle rosapi param responses

bjsowa opened the pull request:
**Public API Changes**
None


**Description**
Thanks to RobotWebTools/rosbridge_suite#1001, since ROS 2 Jazzy, rosapi param services return the `successful` bool and fill the reason for failure if the operation does not succeed. 

This PR adds the logic for handling these additional fields.

bjsowa committed:
Handle rosapi param responses

EzraBrooks commented:
was that backported to humble? roslibjs only has one release channel so I don't think we should break API here if not until Humble is EOL.

bjsowa commented:
> was that backported to humble? roslibjs only has one release channel so I don't think we should break API here if not until Humble is EOL.

It was not backported but this should still be compatible with humble

bjsowa committed:
Merge branch 'develop' into feat/check-param-response

EzraBrooks reviewed (commented):
agree with copilot about not duplicating the code three places, otherwise LGTM. I only skimmed it the first time and didn't notice that you wrote it in a back-compatible way by checking for the presence of the `successful` member.

maybe we need to make `failedCallback` mandatory in a future version, otherwise if someone promisifies this callback-style function this'll just hang forever?

bjsowa committed:
Simplify checking result, call normal callback when failedCallback not provided

bjsowa committed:
Update failedCallback description

bjsowa commented:
@EzraBrooks I simplified it a bit by skipping the "undefined check" (`result.successful === false` will return false if `successful` is undefined) and also changed it so that the normal callback is still called if result is unsuccessful and failedCallback is not provided to keep the old behavior.  Let me know what you think

bjsowa committed:
Merge branch 'develop' into feat/check-param-response

bjsowa committed:
Merge branch 'develop' into feat/check-param-response

bjsowa committed:
Merge branch 'develop' into feat/check-param-response

bjsowa commented:
@EzraBrooks Can we merge this?

bjsowa committed:
Merge branch 'develop' into feat/check-param-response
